// fp-ts-contrib/es6/ReaderTask.js
/**
 * @since 0.1.27
 */
import * as R from 'fp-ts/es6/Reader';
import * as RT from 'fp-ts/es6/ReaderTask';
import * as T from 'fp-ts/es6/Task';
// -------------------------------------------------------------------------------------
// natural transformations
// -------------------------------------------------------------------------------------
/**
 * @category natural transformations
 * @since 0.1.27
 */
// TODO use NaturalTransformation<RIO.URI, RT.URI> when fp-ts version >= 2.11.0
export var fromReaderIO = R.map(T.fromIO);
// -------------------------------------------------------------------------------------
// combinators
// -------------------------------------------------------------------------------------
/**
 * @category combinators
 * @since 0.1.27
 */
export var fromReaderIOK = function (f) { return function () {
    var a = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        a[_i] = arguments[_i];
    }
    return fromReaderIO(f.apply(void 0, a));
}; };
/**
 * Less strict version of [`chainReaderIOK`](#chainreaderiok).
 *
 * @category combinators
 * @since 0.1.27
 */
export var chainReaderIOKW = function (f) { return RT.chainW(fromReaderIOK(f)); };
/**
 * @category combinators
 * @since 0.1.27
 */
export var chainReaderIOK = chainReaderIOKW;
/**
 * Less strict version of [`chainFirstReaderIOK`](#chainfirstreaderiok).
 *
 * @category combinators
 * @since 0.1.27
 */
// TODO use RT.chainFirstW when fp-ts version >= 2.11.0
export var chainFirstReaderIOKW = function (f) { return RT.chainFirst(fromReaderIOK(f)); };
/**
 * @category combinators
 * @since 0.1.27
 */
export var chainFirstReaderIOK = chainFirstReaderIOKW;
