// fp-ts-contrib/lib/index.js
"use strict";
/**
 * @since 0.1.0
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.time = exports.filterA = exports.collectUntil = exports.batchTraverse = exports.regExp = exports.Do = exports.zipper = exports.taskOption = exports.stateTaskEither = exports.stateIO = exports.stateEither = exports.readerIO = exports.list = exports.IOOption = exports.free = exports.semialignNonEmptyArray = exports.semialign = exports.alignRecord = exports.arrayOption = exports.alignArray = exports.align = void 0;
var align = __importStar(require("./Align"));
exports.align = align;
var alignArray = __importStar(require("./Align/Array"));
exports.alignArray = alignArray;
var arrayOption = __importStar(require("./ArrayOption"));
exports.arrayOption = arrayOption;
var alignRecord = __importStar(require("./Align/Record"));
exports.alignRecord = alignRecord;
var semialign = __importStar(require("./Semialign"));
exports.semialign = semialign;
var semialignNonEmptyArray = __importStar(require("./Semialign/NonEmptyArray"));
exports.semialignNonEmptyArray = semialignNonEmptyArray;
var free = __importStar(require("./Free"));
exports.free = free;
var IOOption = __importStar(require("./IOOption"));
exports.IOOption = IOOption;
var list = __importStar(require("./List"));
exports.list = list;
var readerIO = __importStar(require("./ReaderIO"));
exports.readerIO = readerIO;
var stateEither = __importStar(require("./StateEither"));
exports.stateEither = stateEither;
var stateIO = __importStar(require("./StateIO"));
exports.stateIO = stateIO;
var stateTaskEither = __importStar(require("./StateTaskEither"));
exports.stateTaskEither = stateTaskEither;
var taskOption = __importStar(require("./TaskOption"));
exports.taskOption = taskOption;
var zipper = __importStar(require("./Zipper"));
exports.zipper = zipper;
var Do = __importStar(require("./Do"));
exports.Do = Do;
var regExp = __importStar(require("./RegExp"));
exports.regExp = regExp;
var batchTraverse = __importStar(require("./batchTraverse"));
exports.batchTraverse = batchTraverse;
var collectUntil = __importStar(require("./collectUntil"));
exports.collectUntil = collectUntil;
var filterA = __importStar(require("./filterA"));
exports.filterA = filterA;
var time = __importStar(require("./time"));
exports.time = time;
