// fp-ts-contrib/lib/ReaderTask.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.chainFirstReaderIOK = exports.chainFirstReaderIOKW = exports.chainReaderIOK = exports.chainReaderIOKW = exports.fromReaderIOK = exports.fromReaderIO = void 0;
/**
 * @since 0.1.27
 */
var R = __importStar(require("fp-ts/lib/Reader"));
var RT = __importStar(require("fp-ts/lib/ReaderTask"));
var T = __importStar(require("fp-ts/lib/Task"));
// -------------------------------------------------------------------------------------
// natural transformations
// -------------------------------------------------------------------------------------
/**
 * @category natural transformations
 * @since 0.1.27
 */
// TODO use NaturalTransformation<RIO.URI, RT.URI> when fp-ts version >= 2.11.0
exports.fromReaderIO = R.map(T.fromIO);
// -------------------------------------------------------------------------------------
// combinators
// -------------------------------------------------------------------------------------
/**
 * @category combinators
 * @since 0.1.27
 */
var fromReaderIOK = function (f) { return function () {
    var a = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        a[_i] = arguments[_i];
    }
    return (0, exports.fromReaderIO)(f.apply(void 0, a));
}; };
exports.fromReaderIOK = fromReaderIOK;
/**
 * Less strict version of [`chainReaderIOK`](#chainreaderiok).
 *
 * @category combinators
 * @since 0.1.27
 */
var chainReaderIOKW = function (f) { return RT.chainW((0, exports.fromReaderIOK)(f)); };
exports.chainReaderIOKW = chainReaderIOKW;
/**
 * @category combinators
 * @since 0.1.27
 */
exports.chainReaderIOK = exports.chainReaderIOKW;
/**
 * Less strict version of [`chainFirstReaderIOK`](#chainfirstreaderiok).
 *
 * @category combinators
 * @since 0.1.27
 */
// TODO use RT.chainFirstW when fp-ts version >= 2.11.0
var chainFirstReaderIOKW = function (f) { return RT.chainFirst((0, exports.fromReaderIOK)(f)); };
exports.chainFirstReaderIOKW = chainFirstReaderIOKW;
/**
 * @category combinators
 * @since 0.1.27
 */
exports.chainFirstReaderIOK = exports.chainFirstReaderIOKW;
