// fp-ts-contrib/lib/ReaderTaskEither.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.chainFirstReaderIOK = exports.chainFirstReaderIOKW = exports.chainReaderIOK = exports.chainReaderIOKW = exports.fromReaderIOK = exports.rightReaderIO = exports.leftReaderIO = void 0;
/**
 * @since 0.1.27
 */
var function_1 = require("fp-ts/lib/function");
var RTE = __importStar(require("fp-ts/lib/ReaderTaskEither"));
var TE = __importStar(require("fp-ts/lib/TaskEither"));
// -------------------------------------------------------------------------------------
// constructors
// -------------------------------------------------------------------------------------
/**
 * @category constructors
 * @since 0.1.27
 */
var leftReaderIO = function (me) {
    return (0, function_1.flow)(me, TE.leftIO);
};
exports.leftReaderIO = leftReaderIO;
/**
 * @category constructors
 * @since 0.1.27
 */
var rightReaderIO = function (ma) {
    return (0, function_1.flow)(ma, TE.rightIO);
};
exports.rightReaderIO = rightReaderIO;
// -------------------------------------------------------------------------------------
// combinators
// -------------------------------------------------------------------------------------
/**
 * @category combinators
 * @since 0.1.27
 */
var fromReaderIOK = function (f) { return function () {
    var a = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        a[_i] = arguments[_i];
    }
    return (0, exports.rightReaderIO)(f.apply(void 0, a));
}; };
exports.fromReaderIOK = fromReaderIOK;
/**
 * Less strict version of [`chainReaderIOK`](#chainreaderiok).
 *
 * @category combinators
 * @since 0.1.27
 */
var chainReaderIOKW = function (f) {
    return RTE.chainW((0, exports.fromReaderIOK)(f));
};
exports.chainReaderIOKW = chainReaderIOKW;
/**
 * @category combinators
 * @since 0.1.27
 */
exports.chainReaderIOK = exports.chainReaderIOKW;
/**
 * Less strict version of [`chainFirstReaderIOK`](#chainfirstreaderiok).
 *
 * @category combinators
 * @since 0.1.27
 */
var chainFirstReaderIOKW = function (f) {
    return RTE.chainFirstW((0, exports.fromReaderIOK)(f));
};
exports.chainFirstReaderIOKW = chainFirstReaderIOKW;
/**
 * @category combinators
 * @since 0.1.27
 */
exports.chainFirstReaderIOK = exports.chainFirstReaderIOKW;
