// @imgix/gatsby/dist/modules/gatsby-plugin/createImgixUrlFieldConfig.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createImgixUrlSchemaFieldConfig = exports.createImgixUrlFieldConfig = void 0;
var pipeable_1 = require("fp-ts/pipeable");
var T = __importStar(require("fp-ts/Task"));
var TE = __importStar(require("fp-ts/TaskEither"));
var ramda_1 = require("ramda");
var createExternalHelper_1 = require("../../common/createExternalHelper");
var utils_1 = require("../../common/utils");
var graphqlTypes_1 = require("./graphqlTypes");
var createImgixUrlFieldConfig = function (_a) {
    var imgixClient = _a.imgixClient, resolveUrl = _a.resolveUrl, defaultParams = _a.defaultParams, paramsInputType = _a.paramsInputType;
    return ({
        type: graphqlTypes_1.gatsbySourceImgixUrlFieldType,
        description: 'A plain imgix URL with the URL and params applied.',
        args: {
            imgixParams: {
                type: paramsInputType,
                defaultValue: {},
            },
        },
        resolve: function (rootValue, args) {
            return (0, pipeable_1.pipe)(rootValue, (0, utils_1.resolveUrlFromSourceData)(resolveUrl), TE.map(function (url) {
                var _a;
                return imgixClient.buildURL(url, (0, ramda_1.mergeRight)(defaultParams !== null && defaultParams !== void 0 ? defaultParams : {}, (0, graphqlTypes_1.unTransformParams)((_a = args.imgixParams) !== null && _a !== void 0 ? _a : {})));
            }), TE.getOrElse(function () { return T.of(undefined); }))();
        },
    });
};
exports.createImgixUrlFieldConfig = createImgixUrlFieldConfig;
exports.createImgixUrlSchemaFieldConfig = (0, createExternalHelper_1.createExternalHelper)(exports.createImgixUrlFieldConfig);
